Wrong multi-column sort if one of the columns is constant string

```
>>> DT = dt.Frame(A=['abcd'] * 100, B=[random.choice(['a', 'b']) for _ in range(100)], C=[random.choice(['ads', 'adfv', 'adfv']) for  _  in range(100)])
>>> DT[:, dt.count(), dt.by(f.A, f.B, f.C)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Assertion 'nradixes > 0' failed in src/core/sort.cc, line 566

>>> DT[:, dt.count(), dt.by(f.A, f.B)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: Assertion 'groupby_.last_offset() == nr' failed in src/core/expr/eval_context.cc, line 288
```
This bug likely appeared after #3135 

_Originally posted by:_ @arnocandel

--- a/src/core/sort.cc
+++ b/src/core/sort.cc
@@ -1143,7 +1143,7 @@ class SortContext {
       auto rrmap = std::unique_ptr<radix_range[]>(new radix_range[nradixes]);
       radix_range* rrmap_ptr = rrmap.get();
       _fill_rrmap_from_histogram(rrmap_ptr);
-      bool retry = _radix_recurse<make_groups>(rrmap_ptr);
+      bool retry = _radix_recurse<make_groups>(rrmap_ptr, true);
       if (retry) goto start;
       nsigbits = _nsigbits;
     }
@@ -1204,7 +1204,7 @@ class SortContext {
    * radix range, our job will be complete.
    */
   template <bool make_groups>
-  bool _radix_recurse(radix_range* rrmap) {
+  bool _radix_recurse(radix_range* rrmap, bool allow_retry = false) {
     xassert(x && xx && o && next_o);
     // Save some of the variables in SortContext that we will be modifying
     // in order to perform the recursion.
@@ -1259,7 +1259,7 @@ class SortContext {
         // create a very deep chain of recursion which will result in a stack
         // overflow and a crash. With this return status we're effectively
         // doing a manual tail-call elimination. See issue #3134.
-        if (sz == n && off == 0 && is_string) {
+        if (sz == n && off == 0 && is_string && allow_retry) {
           return true;
         }
         elemsize = _elemsize;

--- a/tests/ijby/test-sort.py
+++ b/tests/ijby/test-sort.py
@@ -1161,3 +1161,26 @@ def test_sort_long_nnearly_identical_strings():
     src = ["o" * 20000 + str(i % 10) for i in range(1000)]
     DT = dt.Frame(src).sort(0)
     assert_equals(DT, dt.Frame(sorted(src)))
+
+
+def test_sort_multicolumn1():
+    # See issue #3141
+    DT = dt.Frame(A=[111] * 100,
+                  B=['a', 'b'] * 50,
+                  C=['ads', 'adfv', 'adfv', 'adsfv'] * 25)
+    RES1 = DT[:, dt.count(), dt.by(f.A, f.B, f.C)]
+    assert_equals(RES1,
+        dt.Frame(A=[111, 111, 111, 111],
+                 B=['a', 'a', 'b', 'b'],
+                 C=['adfv', 'ads', 'adfv', 'adsfv'],
+                 count=[25, 25, 25, 25]/dt.int64))
+    RES2 = DT[:, dt.count(), dt.by(f.B, f.C)]
+    assert_equals(RES2,
+        dt.Frame(B=['a', 'a', 'b', 'b'],
+                 C=['adfv', 'ads', 'adfv', 'adsfv'],
+                 count=[25, 25, 25, 25]/dt.int64))
+    RES3 = DT[:, dt.count(), dt.by(f.A, f.B)]
+    assert_equals(RES3,
+        dt.Frame(A=[111, 111],
+                 B=['a', 'b'],
+                 count=[50, 50]/dt.int64))

